docs/DifferentiableProgrammingImplementation.md

```diff
@@ -0,0 +1,1012 @@
+| `y = tuple_extract x, <n>` | `tan[y] = tuple_extract tan[x], <n’>` | `adj[x] += tuple (0, ..., adj[y], ..., 0)` [(code)](https://github.com/apple/swift/blob/8b7ab1143e260d7bb1db3b98e24f7fe28dc7f0f0/lib/SILOptimizer/Mandatory/Differentiation.cpp#L5298)
+| `y = tuple_element_addr x, <n>` | `tan[y] = tuple_element_addr tan[x], <n’>` | No generated code. `adj[y] = tuple_element_addr adj[x], <n’>` [(code)](https://github.com/apple/swift/blob/8b7ab1143e260d7bb1db3b98e24f7fe28dc7f0f0/lib/SILOptimizer/Mandatory/Differentiation.cpp#L4169)
+
+In general, tangent transformation rules are simpler because the tangent transformation does not involve control flow reversal and because many instructions themselves are linear.
+
+### Modules and access levels
+
+Swift organizes code into [modules](https://github.com/apple/swift/blob/master/docs/Modules.rst). Modules enforce access controls on what code can be used outside of them.
+
+Differentiation is designed with modules and access levels in mind. `@differentiable` declaration attributes act like a “differentiability contract”: declarations marked with the attribute can be differentiated from other modules. This is because the differentiation transform is guaranteed to fill in all SIL differentiability witnesses with derivatives functions. Conversely, declarations not marked with `@differentiable` cannot be differentiated from other modules because they do not have registered derivative functions and are essentially opaque (their bodies may not be exposed).
+
+This design makes differentiation modular. The compiler does not need access to the bodies of cross-module declarations in order to differentiate them; it simply looks up their registered derivatives.
+
+### Pseudocode
+
+> TODO: Hyperlink pseudocode lines to actual code.
+
+Here is the pseudocode of the main logic of the differentiation transform.
+
+- For all differentiability witnesses in the current SIL module:
+    - If the witness is missing the JVP function, generate a JVP function and fill it in.
+    - If the witness is missing the VJP function, generate a VJP function and fill it in.
+- Add all `differentiable_function` instructions from the current SIL module to a worklist.
+- While the `differentiable_function` worklist is not empty, pop the next one.
+    - If `differentiable_function` has JVP and VJP values, do nothing. It is already canonical.
+    - If `differentiable_function` is missing JVP and VJP:
+        - If the `differentiable_function`’s original function operand is an `differentiable_function_extract [original]` instruction, get the operand of that instruction. Do `differentiable_function_extract [jvp/vjp]` instruction from that operand to get the JVP/VJP. Continue.
+        - Otherwise, get the “original function reference” (`function_ref`, `witness_method`, or `class_method` instruction) underlying the `differentiable_function`’s original function operand.
```

Comment:
```suggestion
        - Otherwise, get the "original function reference" (`function_ref`, `witness_method`, or `class_method` instruction) underlying the `differentiable_function`’s original function operand.
```